src/main/java/nextstep/qna/service/QnAService.java

```diff
@@ -2,48 +2,45 @@
-import nextstep.qna.domain.*;
+import nextstep.qna.domain.answer.Answer;
+import nextstep.qna.domain.answer.Answers;
+import nextstep.qna.domain.deleteHistory.DeleteHistory;
+import nextstep.qna.domain.question.Question;
+import nextstep.qna.domain.answer.AnswerRepository;
+import nextstep.qna.domain.question.QuestionRepository;
+import nextstep.qna.domain.deleteHistory.type.ContentType;
 import nextstep.users.domain.NsUser;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
 
 import javax.annotation.Resource;
-import java.time.LocalDateTime;
 import java.util.ArrayList;
 import java.util.List;
 
 @Service("qnaService")
 public class QnAService {
-    @Resource(name = "questionRepository")
-    private QuestionRepository questionRepository;
-
-    @Resource(name = "answerRepository")
-    private AnswerRepository answerRepository;
-
-    @Resource(name = "deleteHistoryService")
-    private DeleteHistoryService deleteHistoryService;
-
-    @Transactional
-    public void deleteQuestion(NsUser loginUser, long questionId) throws CannotDeleteException {
-        Question question = questionRepository.findById(questionId).orElseThrow(NotFoundException::new);
-        if (!question.isOwner(loginUser)) {
-            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
-        }
-
-        List<Answer> answers = question.getAnswers();
-        for (Answer answer : answers) {
-            if (!answer.isOwner(loginUser)) {
-                throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
-            }
-        }
-
-        List<DeleteHistory> deleteHistories = new ArrayList<>();
-        question.setDeleted(true);
-        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, questionId, question.getWriter(), LocalDateTime.now()));
-        for (Answer answer : answers) {
-            answer.setDeleted(true);
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
-    }
+  @Resource(name = "questionRepository")
+  private QuestionRepository questionRepository;
+
+  @Resource(name = "answerRepository")
+  private AnswerRepository answerRepository;
+
+  @Resource(name = "deleteHistoryService")
+  private DeleteHistoryService deleteHistoryService;
+
+  @Transactional
+  public void deleteQuestion(NsUser loginUser, long questionId) throws CannotDeleteException {
+
+    // 조회
+    Question question = questionRepository.findById(questionId).orElseThrow(NotFoundException::new);
+    Answers answers = answerRepository.findByQuestion(questionId);
+
+    // 삭제
+    question.delete(loginUser, answers);
+    answers.deleteAll(question.getWriter());
+
+    // 삭제 히스토리 관리
+    List<DeleteHistory> deleteHistories = DeleteHistory.makeDeleteHistories(questionId, question.getWriter(), answers);
+    deleteHistoryService.saveAll(deleteHistories);
```

Comment:
기존 코드에서 리팩토링 잘 해주신 것 같아요. 그러나 문제가 발생할 수 있는 여지가 남아있습니다.
바로 서비스 레이어에서 삭제 히스토리 관리를 직접 해줘야하는 점인데요
delete 메소드 호출 후,
히스토리 객체를 생성하는 것을 놓친다면 삭제는 되나 이력이 저장되지 않는 문제가 발생할 수 있죠

```suggestion

    List<DeleteHistory> deleteHistories = question.delete(loginUser);
    deleteHistoryService.saveAll(deleteHistories);
```
위 처럼 question 에게 delete 메시지를 보내면 `DeleteHistory` 컬렉션을 반환하도록 변경해보면 어떨까요?

Reply:
말씀주셔서 감사합니다!
요 부분 고민했었는데요!

질문 삭제
답변 삭제
삭제에 대한 히스토리 쌓을거야~
를 큰 흐름으로 보여주는 게 좋겠다라고 생각하고 구현했었는데요!


delete 메소드 호출 후,
히스토리 객체를 생성하는 것을 놓친다면 삭제는 되나 이력이 저장되지 않는 문제
=> 요 문제 크리티컬 하고

DeleteHistory 컬렉션을 반환한다면,
삭제하고, 히스토리가 쌓였구나!라고 이해할 수 있어서
말씀주신 방향이 더 좋은 것 같습니다 감사해요~~~